Author: Upstream
Description: Combination of 3 patches for unreleased Expect 5.45.1.
 1) A patch sent in by Ogawa Hirofumi. The patch fixes a problem when
    talking a tty where the writer has died. Some operating systems
    report the condition as EIO with nothing read, while this
    actually an EOF. Without the patch the returned data is
    incomplete due to the error reported immediately and dropping
    data in buffers.
 2) A patch by Michael Cleverly fixing a problem with the
    iteration over the expect channel list where the loop code may
    modify the list, breaking the iterator.
 3) A patch by Michael Cleverly fixing problem with an insufficient
    test for a lost channel in exp_background_channelhandler().
Last-Modified: Fri, 23 Oct 2015 11:24:33 +0300

--- a/expect.c
+++ b/expect.c
@@ -2495,8 +2495,12 @@
      */
 
     /* First check that the esPtr is even still valid! */
-    /* This ought to be sufficient. */
-    if (0 == Tcl_GetChannel(interp,backup,(int *)0)) {
+    /* 
+     * It isn't sufficient to just check that 'Tcl_GetChannel' still knows about
+     * backup since it is possible that esPtr was lost in the background AND
+     * another process spawned and reassigned the same name. 
+     */
+    if (!expChannelStillAlive(esPtr, backup)) {
       expDiagLog("expect channel %s lost in background handler\n",backup);
       return;
     }
--- a/exp_chan.c
+++ b/exp_chan.c
@@ -218,6 +218,11 @@
 
     bytesRead = read(esPtr->fdin, buf, (size_t) toRead);
     /*printf("ExpInputProc: read(%d,,) = %d\r\n",esPtr->fdin,bytesRead);*/
+
+    /* Emulate EOF on tty for tcl */
+    if ((bytesRead == -1) && (errno == EIO) && isatty(esPtr->fdin)) {
+	bytesRead = 0;
+    }
     if (bytesRead > -1) {
 	/* strip parity if requested */
 	if (esPtr->parity == 0) {
@@ -448,6 +453,34 @@
     ThreadSpecificData *tsdPtr = TCL_TSD_INIT(&dataKey);
     return tsdPtr->channelCount;
 }
+
+int
+expChannelStillAlive(esBackupPtr, backupName)
+     ExpState *esBackupPtr;
+     char *backupName;
+{
+    ThreadSpecificData *tsdPtr = TCL_TSD_INIT(&dataKey);
+    ExpState *esPtr;
+
+    /* 
+     * This utility function is called from 'exp_background_channelhandler'
+     * and checks to make sure that backupName can still be found in the 
+     * channels linked list at the same address as before.
+     *
+     * If it can't be (or if the memory address has changed) then it
+     * means that it was lost in the background (and possibly another
+     * channel was opened and reassigned the same name).
+     */
+
+    for (esPtr = tsdPtr->firstExpPtr; esPtr; esPtr = esPtr->nextPtr) {
+        if (0 == strcmp(esPtr->name, backupName)) 
+            return (esPtr == esBackupPtr);
+    }
+        
+    /* not found; must have been lost in the background */
+    return 0;
+}
+
 #if 0 /* Converted to macros */
 int
 expSizeGet(esPtr)
@@ -603,12 +636,38 @@
 {
     ThreadSpecificData *tsdPtr = TCL_TSD_INIT(&dataKey);
     ExpState *esPtr;
+    ExpState *esNextPtr;
+    ExpState *esPriorPtr = 0;
 
     /* kick off any that already have input waiting */
-    for (esPtr = tsdPtr->firstExpPtr;esPtr;esPtr = esPtr->nextPtr) {
+    for (esPtr = tsdPtr->firstExpPtr;esPtr; esPriorPtr = esPtr, esPtr = esPtr->nextPtr) {
 	/* is bg_interp the best way to check if armed? */
 	if (esPtr->bg_interp && !expSizeZero(esPtr)) {
+            /* 
+             * We save the nextPtr in a local variable before calling
+             * 'exp_background_channelhandler' since in some cases
+             * 'expStateFree' could end up getting called before it
+             * returns, leading to a likely segfault on the next
+             * interaction through the for loop.
+             */
+            esNextPtr = esPtr->nextPtr;
 	    exp_background_channelhandler((ClientData)esPtr,0);
+            if (esNextPtr != esPtr->nextPtr) {
+                /* 
+                 * 'expStateFree' must have been called from
+                 * underneath us so we know that esPtr->nextPtr is
+                 * invalid.  However, it is possible that either the
+                 * original nextPtr and/or the priorPtr have been
+                 * freed too.  If the esPriorPtr->nextPtr is now
+                 * esNextPtr it seems safe to proceed.  Otherwise we
+                 * break and end early for safety.
+                 */
+                if (esPriorPtr && esPriorPtr->nextPtr == esNextPtr) {
+                    esPtr = esPriorPtr;
+                } else {
+                    break; /* maybe set esPtr = tsdPtr->firstExpPtr again? */
+                }
+            }
 	}
     }
 }
--- a/exp_command.h
+++ b/exp_command.h
@@ -248,6 +248,7 @@
 EXTERN void		exp_ecmd_remove_state_direct_and_indirect _ANSI_ARGS_((Tcl_Interp *interp,ExpState *esPtr));
 EXTERN void		expChannelInit _ANSI_ARGS_((void));
 EXTERN int		expChannelCountGet _ANSI_ARGS_((void));
+EXTERN int              expChannelStillAlive _ANSI_ARGS_((ExpState *, char *));
 
 EXTERN int		exp_tcl2_returnvalue _ANSI_ARGS_((int));
 EXTERN int		exp_2tcl_returnvalue _ANSI_ARGS_((int));
